core/bin/dust_api.rs

```diff
@@ -1213,12 +1215,96 @@ async fn data_sources_documents_update_tags(
             }),
         );
     }
+
+    match load_data_source(&state, &project, &data_source_id).await {
+        Ok(ds) => match ds
+            .update_tags(
+                state.store.clone(),
+                state.qdrant_client.clone(),
+                document_id,
+                add_tags_set.into_iter().collect(),
+                remove_tags_set.into_iter().collect(),
+            )
+            .await
+        {
+            Err(e) => server_error_response("Failed to update document tags", e),
+            Ok(_) => datasource_ok_response(ds),
+        },
+        Err(api_error) => api_error,
+    }
+}
+
+/// Update parents of a document in a data source.
+
+#[derive(serde::Deserialize)]
+struct DataSourcesDocumentsUpdateParentsPayload {
+    parents: Vec<String>,
+}
+
+async fn data_sources_documents_update_parents(
+    extract::Path((project_id, data_source_id, document_id)): extract::Path<(i64, String, String)>,
+    extract::Json(payload): extract::Json<DataSourcesDocumentsUpdateParentsPayload>,
+    extract::Extension(state): extract::Extension<Arc<APIState>>,
+) -> (StatusCode, Json<APIResponse>) {
+    let project = project::Project::new_from_id(project_id);
+
+    match load_data_source(&state, &project, &data_source_id).await {
+        Ok(ds) => match ds
+            .update_parents(
+                state.store.clone(),
+                state.qdrant_client.clone(),
+                document_id,
+                payload.parents,
+            )
+            .await
+        {
+            Err(e) => server_error_response("Failed to update document parents", e),
+            Ok(_) => datasource_ok_response(ds),
+        },
+        Err(api_error) => api_error,
```

Comment:
I believe this is incorrect we don't want to return the raw error but a well formatted error and I don't believe that object is one? Where is the None case also which should create a NOT_FOUND response?

Reply:
Ah I see you dried that one also. Then you need to dry the whole file, assuming we do behave the same way everywhere which I'm not convinced of.

I strongly believe that seeing all the errors and all the response of an API endpoint in its implementation has a great value for understanding the code rapidly. This is another abstraction that I find more diverting than useful.

Reply:
It does not really load a data source, it loads a data source but return a very special error for the API... so really it should be load_data_source_or_error_with_api_format which I would argue is not the most superb abstraction

Reply:
I'll revert it; I can make my counterpoint and we can go on the debate over coffee or lunch :)